### DISCUSSION TITLE
Transfer of "Cake.DotNetLocalTools.Module" and "Cake.GitHubReleases" into the Cake Contrib organization

### DISCUSSION THREAD

**Discussion by ap0llo** [cake-build/cake#3618] [Extension requests]
Hi,

I recently published two Cake extensions to nuget.org:  [Cake.GitHubReleases](https://www.nuget.org/packages/Cake.GitHubReleases/) and [Cake.DotNetLocalTools.Module](https://www.nuget.org/packages/Cake.DotNetLocalTools.Module).

The GitHub repositories are currently both under my personal account:
- https://github.com/ap0llo/Cake.GitHubReleases
- https://github.com/ap0llo/Cake.DotNetLocalTools.Module

I would like to transfer them into the Cake Contrib organization (I would continue to maintain these projects).

However it's not a 100% clear to me what that means:

- The documentation mentions an AppVeyor project that will be used for projects after the transfer. Is it expected to use that for CI/CD? I already have a pipeline on Azure DevOps in place.
- Are there any additional conventions or best practices all projects are expected to follow?
- Are there any requirements to use a specific license? I usually use the MIT license for my side-projects.

**Comment by gep13** [comment]
There is no requirement to use AppVeyor, no, this is just an option if you don't already have something set up.

Have you had a chance to read through this:

https://github.com/cake-contrib/Home/blob/master/Guidelines.md

There are a couple of NuGet packages that might be useful to you:

* https://www.nuget.org/packages/CakeContrib.Guidelines/
* https://www.nuget.org/packages/Cake.Addin.Analyzer/
 
MIT license is perfect, so no issues there.

**Comment by gitfool** [comment]
Support for dotnet local tools looks particularly interesting to me and could potentially fix https://github.com/cake-build/cake/issues/3209#issuecomment-787210882.

Also, given the dotnet tool module was moved into Cake core, could this be moved or better yet directly added there too?

**Reply by gep13** [reply]
I believe that this is something that we have thought about, but no decision has been made on how to move forward, and we would need to work with @ap0llo to decide what is best.

**Reply by pascalberger** [reply]
We quickly talked about this on our last meeting. Currently our focus is on getting 2.0 done, but it would absolutely make sense to have support for local tools as part of core Cake functionality. 

I don't think anybody had looked at the module in detail, but if you or @ap0llo see a possibility to bring this into Cake to solve #3209, we're happy to discuss possibilities.

**Reply by ap0llo** [reply]
I'd be open to eventually bringing this into Cake core, however I'm not sure if the module provides an exact solution for what's described in #3209.

The module does not exactly use local tools (as in running them through `dotnet tool run ...`) but rather reads out the tools references from a `dotnet-tools.json` file and passes them to `DotNetToolPackageInstaller` that already exists in Cake, which will then install the tool sane as if it was specified using the `#tool` directive or `InstallTool()` - so the tool manifest becomes an alternative way to tell Cake what to install.

You mentioned this would be something for post-2.0. What time-frame would that be?

**Reply by pascalberger** [reply]
> You mentioned this would be something for post-2.0. What time-frame would that be?

To get something like support for .NET local tools in, someone needs to champion it. Our main focus is now getting stuff for 2.0 done and probably nobody from the core maintainers has much resources to dive into this. But if you, or someone else from community, wants to push this and sketch out ideas how it can be implemented in core, so that we can discuss them, this would be appreciated. 

Regarding time-frame: We do feature scoped and not time scoped release. Current plan is to release 2.0 in November.

**Comment by pascalberger** [comment]
> * The documentation mentions an AppVeyor project that will be used for projects after the transfer. Is it expected to use that for CI/CD? I already have a pipeline on Azure DevOps in place.

You're absolutely free to use what you prefer. Currently repos in cake-contrib are using AppVeyor, Azure Pipelines and GitHub actions AFAIK. If you already use Azure Pipeline I can also create a project on the [cake-contrib Azure DevOps tenant](https://dev.azure.com/cake-contrib/) for you.

> * Are there any additional conventions or best practices all projects are expected to follow?

On the website you can find general best practices for [Addins](https://cakebuild.net/docs/extending/addins/best-practices) and [Modules](https://cakebuild.net/docs/extending/modules/best-practices) .

There are also some additional [guidelines for cake-contrib repos](https://github.com/cake-contrib/Home/blob/master/Guidelines.md). Using common infrastructure as suggested helps to get people to contribute, but they are not hard requirements.

**Reply by ap0llo** [reply]
Thanks for your clarifications.

I'd like to move forward with moving "Cake.DotNetLocalTools.Module" into the Cake Contrib org and will have a look into the guidelines and adapt the project over time.

I don't think it is necessary to move the Cake.GitHubReleases repository as I will work on merging that into Cake.GitHub (as you suggested in https://github.com/ap0llo/Cake.GitHubReleases/issues/1)

**Reply by gep13** [reply]
@ap0llo that is great news!

To start the process of moving `Cake.DotNetLocalTools.Module` into the Cake-Contrib Organisation, it needs to be done by someone with Owner permissions of that Organisation.  The easiest way to do this would be for you to transfer the repository to my personal account `gep13`, and from there, I can immediately move it to the Cake-Contrib Organisation, and give you admin permissions to it in there.

How does that sound?

**Reply by ap0llo** [reply]
Sounds good, I'll transfer the repo

**Reply by gep13** [reply]
@ap0llo thank you for doing that!  I have transferred the repository into the Cake-Contrib Organisation, and invited you to the Organisation.  The repository can now be found here:

https://github.com/cake-contrib/Cake.DotNetLocalTools.Module

**Reply by ap0llo** [reply]
Thanks!

**Reply by gep13** [reply]
@ap0llo could you go ahead and mark this discussion as answered? Thanks

**Reply by pascalberger** [reply]
@gep13 Answering is currently only enabled for the `Q & A` categories and not enabled in the `Extension requests` category

**Reply by gep13** [reply]
Ah! Makes sense! Thanks for clarifying, I thought I couldn't see the option because I didn't create the discussion.
